src/overlay/OverlayManagerImpl.cpp

```diff
@@ -1115,4 +1123,165 @@ OverlayManagerImpl::updateFloodRecord(StellarMessage const& oldMsg,
     ZoneScoped;
     mFloodGate.updateRecord(oldMsg, newMsg);
 }
+
+void
+OverlayManagerImpl::demand()
+{
+    ZoneScoped;
+    if (mShuttingDown)
+    {
+        return;
+    }
+    auto const& authenticatedPeers = getAuthenticatedPeers();
+    std::vector<std::pair<NodeID, Peer::pointer>> peers;
+    for (auto const& p : authenticatedPeers)
+    {
+        if (p.second->shouldFloodLazily())
+        {
+            peers.push_back(p);
+        }
+    }
+    stellar::shuffle(peers.begin(), peers.end(), gRandomEngine);
+
+    std::map<NodeID, std::vector<Hash>> demandMap;
+
+    auto now = mApp.getClock().now();
+
+    auto const& cfg = mApp.getConfig();
+    auto remPeers = peers;
+    std::map<NodeID, std::queue<Hash>> nextRetryMap;
+    for (;;)
+    {
+        std::vector<std::pair<NodeID, Peer::pointer>> nextPeers;
+        for (auto& p : remPeers)
+        {
+            if (demandMap[p.first].size() >= cfg.FLOOD_MAX_DEMAND_SIZE)
+            {
+                continue;
+            }
+            Hash code;
+            bool noHashLeft = false;
+            if (mRetryMap[p.first].size() > 0)
+            {
+                code = mRetryMap[p.first].front();
+                mRetryMap[p.first].pop();
+            }
+            else if (p.second->txHashCount() > 0)
+            {
+                code = p.second->popFrontTxHash();
+            }
+            else
+            {
+                noHashLeft = true;
+            }
+            if (!noHashLeft)
+            {
+                auto const& herder = mApp.getHerder();
+
+                if (!herder.isBannedTx(code) && herder.getTx(code) == nullptr)
+                {
+                    // We don't know this txn and this isn't banned.
+                    // Should we pull this txn?
+
+                    // We should demand if and only if:
+                    // (1) We've never demanded this code, OR
+                    // (2) We haven't demanded too many times & it's been
+                    // sufficiently long since the last demand (linear
+                    // back-off).
+                    bool shouldDemand = false;
+                    int numDemanded = mDemandHistory[code].first;
+                    if (numDemanded == 0)
+                    {
+                        // never demanded
+                        shouldDemand = true;
+                    }
+                    else if (numDemanded >= MAX_RETRY_COUNT)
+                    {
+                        shouldDemand = false;
```

Comment:
looks like we still add hash to the retries, which point to missing tests

Reply:
Fixed.